src/base/memory_monitor_sr.cpp

```diff
@@ -0,0 +1,258 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+/*
+ * memory_monitor_sr.cpp - implementation of memory monitoring manager
+ */
+
+#include <stdexcept>
+#include <cstring>
+
+#include "perf.hpp"
+#include "error_manager.h"
+#include "memory_monitor_sr.hpp"
+
+#if defined (SERVER_MODE)
+#include "log_impl.h"
+#endif
+
+
+#ifdef SERVER_MODE
+namespace cubperf
+{
+  memory_monitoring_manager *mmm_Gl = nullptr;
+
+  const char *mmm_subcomponent::get_name ()
+  {
+    return m_subcomp_name;
+  }
+
+  void mmm_subcomponent::add_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_add (size);
+  }
+
+  void mmm_subcomponent::sub_cur_stat (uint64_t size)
+  {
+    m_cur_stat.fetch_sub (size);
+  }
+
+  const uint64_t mmm_subcomponent::get_cur_stat ()
+  {
+    return m_cur_stat.load ();
+  }
+
+  const char *mmm_component::get_name ()
+  {
+    return m_comp_name;
+  }
+
+  void mmm_component::add_stat (uint64_t size, bool init)
+  {
+    return;
+  }
+
+  void mmm_component::sub_stat (uint64_t size, bool init)
+  {
+    return;
+  }
+
+  void mmm_component::add_expand_count ()
+  {
+    m_stat.expand_count++;
+  }
+
+  const MMM_MEM_STAT &mmm_component::get_stat ()
+  {
+    return m_stat;
+  }
+
+  const std::vector<mmm_subcomponent *> &mmm_component::get_subcomp_vec ()
+  {
+    return m_subcomponent;
+  }
+
+  int mmm_component::add_subcomponent (const char *name)
+  {
+    return 0;
+  }
+
+  mmm_module::mmm_module (const char *name, const MMM_COMP_INFO *info)
+  {
+    m_module_name = new char[strlen (name) + 1];
+    strcpy (m_module_name, name);
+
+    /* register component and subcomponent information
+     * add component and subcomponent */
+    int cnt = 0;
+    while (info[cnt].id != MMM_STAT_LAST)
+      {
+	bool comp_skip = false;
+	bool subcomp_skip = false;
+	int comp_idx = mmm_module::MAX_COMP_IDX, subcomp_idx = mmm_module::MAX_COMP_IDX;
+	int i;
+	if (info[cnt].comp_name)
+	  {
+	    for (i = 0; i < m_component.size (); i++)
+	      {
+		if (!strcmp (info[cnt].comp_name, m_component[i]->get_name ()))
+		  {
+		    comp_idx = i;
+		    comp_skip = true;
+		    break;
+		  }
+	      }
+
+	    if (!comp_skip)
+	      {
+		comp_idx = add_component (info[cnt].comp_name);
+	      }
+
+	    if (info[cnt].subcomp_name)
+	      {
+		std::vector<mmm_subcomponent *> subcomp_vec = m_component[comp_idx]->get_subcomp_vec();
+		for (i = 0; i < subcomp_vec.size (); i++)
+		  {
+		    if (!strcmp (info[cnt].subcomp_name, subcomp_vec[i]->get_name ()))
+		      {
+			subcomp_idx = i;
+			subcomp_skip = true;
+			break;
+		      }
+		  }
+
+		if (!subcomp_skip)
+		  {
+		    subcomp_idx = m_component[comp_idx]->add_subcomponent (info[cnt].subcomp_name);
+		  }
+	      }
+	  }
+
+	m_comp_idx_map.push_back (MMM_MAKE_STAT_IDX_MAP (comp_idx, subcomp_idx));
+
+	cnt++;
+      }
+  }
+
+  int mmm_module::aggregate_stats (MEMMON_MODULE_INFO *info)
```

Comment:
```suggestion
  int mmm_module::aggregate_stats (const MEMMON_MODULE_INFO *info)
```
if it's not array, though I don't know how it's used yet. And `const MEMMON_MODULE_INFO&) would be better.

Reply:
Done.